STORM-2225: change spout config to be simpler.

revans2 opened the pull request:
STORM-2228: removed ability to request a single topic go to multiple streams

revans2 commented:
I updated the docs and the examples.  This is a non-backwards compatible change from the 1.x release.  I have some code (not included here) that can maintain most compatibility if we really want to, but because of STORM-2228 I decided it would be better to just not support that at all.  If we want to add that compatibility for 1.x  but deprecated I am OK with that, but would like feedback on it before putting up some kind of a pull request.

d2r commented:
I see some duplicate code and documentation.  Is the idea to move from /external/storm-kafka to /external/storm-kafka-client?

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -60,123 +68,197 @@
+         * The minimum amount of data the broker should return for a fetch request.
+         */
+        public Builder<K,V> setFetchMinBytes(int bytes) {
+            return setProp("fetch.min.bytes", bytes);
```
Nit: Consider using some of the constants in `org.apache.kafka.clients.consumer.ConsumerConfig` instead

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -60,123 +68,197 @@
+         * Only will work with Kafak 0.10.0 and above.
+         */
+        public Builder<K,V> setMaxPoolRecords(int records) {
+            return setProp("max.poll.recordes", records);
```
recordes -> records

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -60,123 +68,197 @@
+        
+        /**
+         * The maximum number of records a poll will return.
+         * Only will work with Kafak 0.10.0 and above.
```
kafak -> kafka, and "will only work" instead of "only will work"

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/NamedSubscription.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public String getTopicsString() {
+        StringBuilder ret = new StringBuilder();
```
Why not use https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.Iterable-?

revans2 replied:
Another 1.8 addition that I didn't know about.

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/RecordTranslator.java:
```diff
@@ -0,0 +1,59 @@
+import org.apache.storm.tuple.Fields;
+
+/**
+ * Translates a Kafka ConsumerRecord into a tuple
```
Consider adding a period at the end here, when shown as javadoc this is going to look weird otherwise.

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/RecordTranslator.java:
```diff
@@ -0,0 +1,59 @@
+ * to different streams.
+ */
+public interface RecordTranslator<K, V> extends Serializable, Function<ConsumerRecord<K, V>, List<Object>> {
+    public static final List<String> DEFAULT_STREAM = Collections.unmodifiableList(Arrays.asList("default"));
```
Collections.singletonList is an easier way to do this

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/RecordTranslator.java:
```diff
@@ -0,0 +1,59 @@
+
+/**
+ * Translates a Kafka ConsumerRecord into a tuple
+ * use {@link KafkaTuple} if you want to route your tuples
```
Nit: I think it's a little clearer to say "Translate to KafkaTuple if..." or "Return a KafkaTuple if..."

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:
```diff
@@ -185,7 +178,9 @@ private long doSeek(TopicPartition tp, OffsetAndMetadata committedOffset) {
 
     private Collection<TopicPartition> toArrayList(final TopicPartition tp) {
-        return new ArrayList<TopicPartition>(1){{add(tp);}};
+        ArrayList<TopicPartition> ret = new ArrayList<>(1);
```
Nit: Unless you need to modify the return value later, you can just return Collections.singleton(tp)

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java:
```diff
@@ -152,7 +152,9 @@ private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata<K, V> lastBa
 
     private Collection<TopicPartition> toArrayList(final TopicPartition tp) {
-        return new ArrayList<TopicPartition>(1){{add(tp);}};
+        ArrayList<TopicPartition> ret = new ArrayList<>(1);
```
Nit: Same as in KafkaSpout

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutManager.java:
```diff
@@ -44,29 +40,30 @@
+                fields = translator.getFieldsFor(stream);
+            } else {
+                if (!fields.equals(translator.getFieldsFor(stream))) {
+                    throw new IllegalArgumentException("Trident Spouts do nut support multiple output Fields");
```
Nit: nut -> not. Also maybe rephrase as "must have the same fields for all streams"

srdo commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/bolt/KafkaBoltTest.java:
```diff
@@ -0,0 +1,91 @@
+                return null;
+            }
+        });
+        KafkaBolt<String, String> bolt = new KafkaBolt<String, String>() {
```
Nit: Can leave off the contents of `<>` on the right hand side in a few places here

revans2 replied:
In this case because I am subclassing KafkaBolt it the compiler(at least in eclipse) actually complains that I am not allowed to do it.

srdo replied:
Right, can't do that. Sorry :)

srdo commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/ByTopicRecordTranslatorTest.java:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void testFiledCollision() {
```
filed -> field

srdo commented:
I like the new configuration design. Backward compatibility is not an issue for us. I'm wondering if enough people have even switched to the new spout yet to make backward compatibility for 1.x a must.

revans2 commented:
@srdo for me backwards compatibility for 1.x is more a question of violating out versioning than anything else.

srdo commented:
Sure, that makes sense. Maybe storm-kafka-client should have been marked unstable since it's fairly new, it's still in a phase where the kinks are being worked out, even if the version number says differently.

revans2 commented on docs/storm-kafka-client.md:
```diff
@@ -1,90 +1,222 @@
+//For topic_2 all events will be emitted to "STREAM_2" as just "key" and "value"
+byTopic.forTopic("topic_2", (r) -> new Values(r.key(), r.value()), new Fields("key", "value"), "STREAM_2");
+
+tp.setSpout("kafka_spout", new KafkaSpout<>(KafkaSpoutConfig.builder("127.0.0.1:" + port, Pattern.compile("topic_.*")).build()), 1);
```
That all depends on the topics defined in the kafka cluster.  If it only has `topic_2` and `topic` defined then yes, but if it has `topic_1` and `topic_2` defined then it will match both of them.

revans2 replied:
I'll try to make the docs clearer

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,194 @@
+                            synchronized (collector) {
+                                if (e != null) {
+                                    collector.reportError(e);
+                                    collector.fail(input);
```
collector could always be called from background threads.  There, however, was a race in the original shuffle grouping that could crash your worker if you didn't synchronize around calls to the collector.  I fixed that a long time ago, but people are still very nervous about it, or are still confused about it.

srdo replied:
Sounds good then, thanks

revans2 commented:
@srdo I think I addressed all of your review comments.

srdo commented:
I think my only remaining nit is the spelling error in KafkaBolt. Thanks @revans2. I am +1 on this.

hmcl commented:
@revans2 I would like to take a look at this before this gets merged in.

revans2 commented:
@srdo as part of backporting this to 1.x I am going to need to make a change to not use Function directly, because it is only in java 8.  So to maintain compatibility between 1.x and 2.x I am going to need to make a few changes in this patch too.

HeartSaVioR commented on external/storm-kafka-client/README.md:
```diff
@@ -1,191 +1,3 @@
-
-## Committer Sponsors
- * Sriharsha Chintalapani ([sriharsha@apache.org](mailto:sriharsha@apache.org))
+Please see [here](../../docs/storm-kafka-client.md)
```
We still need to keep the description because this file will be included to binary dist whereas docs is not.

revans2 commented:
I think I have addressed all of the review comments so far.  I will try to get my 1.x version of the patch up shortly.

srdo commented on external/storm-kafka-client/README.md:
```diff
@@ -1,191 +1,5 @@
+#Storm Kafka integration using the kafka-client jar
 
-Apache Storm Spout implementation to consume data from Apache Kafka versions 0.10 onwards (please see [Apache Kafka Version Compatibility] (#compatibility)). 
+Spouts and Bolts used to interact with Kafka thought the kafka-client library.
```
Nit: "Thought" should be "through"

revans2 commented:
All outstanding review comments should be done now.  This and the 1.x port at #1868 should be ready for a final pass and hopefully being merged in.

revans2 commented:
The test failures are unrelated and are around the integration tests that always seem to fail lately.

revans2 commented:
@srdo @harshach sorry to do this to you, but I just fixed the conflicts with STORM-2236.  Sadly the fastest way I could do it was to revert the original code and implement similar functionality, which is the latest commit.  Could you please take a look at it?  I created a few new Subscription implementations that can do the manual partition management.  It only needed a small change to the Spout to support a timeout.  I will try to look at adding some documentation and also the impact to the trident spout.

hmcl commented on docs/storm-kafka-client.md:
````diff
@@ -1,90 +1,232 @@
+```
+
+#### Wildcard Topics
+Wildcard topics will consume from all topics that are on the configured broakers and match the pattern.  So in the following example
````
... that exist in the brokers list specified and...

hmcl commented on docs/storm-kafka-client.md:
```diff
@@ -1,90 +1,232 @@
+
+#### Wildcard Topics
+Wildcard topics will consume from all topics that are on the configured broakers and match the pattern.  So in the following example
+"topic", "topic_foo" and "topic_bar" will all match the pattern, but "not_my_topic" would not match. 
```
the pattern "topic.*"

hmcl commented on docs/storm-kafka-client.md:
```diff
@@ -1,90 +1,232 @@
+a spout.
+
+`bootstrapServers` is the same as the Kafka Consumer Property "bootstrap.servers".
+`topics` The topics the spout will consume can either be a `Collection` of specific topic names (1 or more) or a regular expression `Pattern` and any
```
... expression, which specifies that any topics ...

hmcl commented on docs/storm-kafka-client.md:
```diff
@@ -1,90 +1,232 @@
+`topics` The topics the spout will consume can either be a `Collection` of specific topic names (1 or more) or a regular expression `Pattern` and any
+topics that match that regular expression will be consumed.
+
+In the case of the Constructors you may also need to specify a key deserializer and a value deserializer.  This is to help make the java generics happy
```
This is to guarantee type safety through the use of Java generics

hmcl commented on docs/storm-kafka-client.md:
```diff
@@ -1,90 +1,232 @@
+ * `UNCOMMITTED_EARLIEST` (DEFAULT) means that the kafka spout polls records from the last committed offset, if any. If no offset has been committed, it behaves as `EARLIEST`.
+ * `UNCOMMITTED_LATEST` means that the kafka spout polls records from the last committed offset, if any. If no offset has been committed, it behaves as `LATEST`.
+
+`setRecordTranslator` allows you to modify how the spout converts a `ConsumerRecord` into a Tuple and which stream that tuple will go to.  By default the "topic",
```
Kafka `Consumer Record` into a Tuple, and which stream that tuple will be published into.

hmcl commented on docs/storm-kafka-client.md:
```diff
@@ -1,90 +1,232 @@
+"partition", "offset", "key", and "value" will be emitted to the "default" stream.  If you want to output entries to different streams based on the topic storm
+provides `ByTopicRecordTranslator`.  See below for more examples on how to use these.
+
+`setProp` can be used to set kafka properties that do not have a convenience method.
```
"Optional" -  should we rename this method to `setKafkaProp` or `setKafkaBrokerProp`  ?

revans2 replied:
We are configuring the kafka consumer, so I would assume that it was implied.

hmcl commented on docs/storm-kafka-client.md:
```diff
@@ -1,90 +1,232 @@
+### Usage Examples
+
+#### Create a Simple Insecure Spout
+The following will consume all events published to "topic" and send them to MyBolt as "topic", "partition", "offset", "key", "value".
```
... MyBolt using the fields "topic" ...

hmcl commented on docs/storm-kafka-client.md:
````diff
@@ -1,90 +1,232 @@
+```
+
+#### Wildcard Topics
+Wildcard topics will consume from all topics that exist in the specified broakers list and match the pattern.  So in the following example
````
typo - brokers

revans2 commented:
The test failures are unrelated

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,194 @@
+    private OutputCollector collector;
+    private TupleToKafkaMapper<K,V> mapper;
+    private KafkaTopicSelector topicSelector;
+    private Properties boltSpecfiedProperties = new Properties();
```
nit: `boltSpecifiedProps` or `boltSpecificProps`

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+        return this;
+    }
+
+    public KafkaBolt<K, V> withTopicSelector(String topic) {
```
`withDefaultTopicSelector` - this would make clear to the API user the difference between the overloaded methods

revans2 replied:
I don't want to rename public facing APIs right now, because this is a copy of what is in external/storm-kafka.  The other one is deprecated, but I want to maintain compatibility if possible.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+        return this;
+    }
+
+    public KafkaBolt<K,V> withProducerProperties(Properties producerProperties) {
```
Is this method intended to set KafkaProducerProperties only? If so, wouldn't it be beneficial to call this method `withKafkaProducerProps`?

revans2 replied:
This is "moving" existing code so I want to maintain compatibility if possible.  But we are doing a copy + deprecation because the two libraries are compiled with different versions of Kafka so combining the two libraries in a single topology is difficult in some cases.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+            if(stormConf.containsKey(TOPIC)) {
+                this.topicSelector = new DefaultTopicSelector((String) stormConf.get(TOPIC));
+            } else {
+                throw new IllegalArgumentException("topic should be specified in bolt's configuration");
```
`IllegalStateException` ?

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+        //for backward compatibility.
+        if(mapper == null) {
+            this.mapper = new FieldNameBasedTupleToKafkaMapper<K,V>();
```
perhaps log here
`LOG.info("Mapper not specified. Setting default mapper to {},  FieldNameBasedTupleToKafkaMapper.class.getSimpleName());`

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+        //for backward compatibility.
+        if(topicSelector == null) {
+            if(stormConf.containsKey(TOPIC)) {
+                this.topicSelector = new DefaultTopicSelector((String) stormConf.get(TOPIC));
```
perhaps log here
LOG.info("TopicSelector not specified. Using [{}] for topic [{}] specified in bolt configuration, DefaultTopicSelector.class.getSimpleName(), stormConf.get(TOPIC));

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+    /**
+     * Intended to be overridden for tests.  Make the producer from props
+     */
+    protected KafkaProducer<K, V> mkProducer(Properties props) {
```
`newProducer` ?

revans2 replied:
backwards compatibility

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+    }
+    
+    /**
+     * Intended to be overridden for tests.  Make the producer from props
```
Create producer with props.

revans2 replied:
backwards compatibility

revans2 replied:
No sorry wrong place for that comment.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+    private TupleToKafkaMapper<K,V> mapper;
+    private KafkaTopicSelector topicSelector;
+    private Properties boltSpecfiedProperties = new Properties();
+    private boolean fireAndForget = false;
```
Can you please add a comment saying to look at the comments in the set method? I missed those comments at first.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+    private KafkaTopicSelector topicSelector;
+    private Properties boltSpecfiedProperties = new Properties();
+    private boolean fireAndForget = false;
+    private boolean async = true;
```
Can you please add a comment saying to look at the comments in the set method? I missed those comments at first.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    /**
+     * If set to true the bold will assume that sending a message to kafka will succeed and will ack
```
bolt

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -0,0 +1,196 @@
+    /**
+     * If set to true the bold will assume that sending a message to kafka will succeed and will ack
+     * the tuple as soon as it has handed the message off to the producer API
+     * if false (the default) the message will be acked after it we successfully sent to kafka or
```
it was successfully

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/Subscription.java:
```diff
@@ -0,0 +1,62 @@
+    }
+    
+    /**
+     * @return a string representing the subscribed topics.
```
Can this string be in any format? It seems to only be used in getComponentConfiguration, not really sure where that's being read.

revans2 replied:
I didn't know either, but the only place I can see it used is in logging so that is what I did.

srdo commented on external/storm-kafka/src/jvm/org/apache/storm/kafka/bolt/KafkaBolt.java:
```diff
@@ -51,7 +51,9 @@
  * It works for sending tuples to older Kafka version (0.8.1).
+ * @deprecated Please use the KafkaBolt in storm-kafka-client
  */
+@Deprecated
```
Maybe it would be fine to just remove it here immediately and deprecate it in 1.1.0?

revans2 replied:
I would rather do that in a follow on JIRA.  Most of the code is the same between the two so getting the rework from here to make there is a lot simpler if I can do a cherry pick

revans2 replied:
I filed https://issues.apache.org/jira/browse/STORM-2317 to do this and I assigned it to myself.

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -61,129 +66,244 @@
+         * The maximum number of records a poll will return.
+         * Will only work with Kafka 0.10.0 and above.
+         */
+        public Builder<K,V> setMaxPoolRecords(int records) {
```
maxPool -> maxPoll

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutManager.java:
```diff
@@ -44,29 +43,30 @@
-        kafkaSpoutStreams = kafkaSpoutConfig.getKafkaSpoutStreams();
-        tuplesBuilder = kafkaSpoutConfig.getTuplesBuilder();
+        RecordTranslator<K, V> translator = kafkaSpoutConfig.getTranslator();
+        Fields fields = null;
```
I would put all of this logic in a method called `setFields`. However, this is going to make this class hard to unit test. So we should even create a static factory method just wrapping this logic, and the constructor with two parameters, `KafkaTridentSpoutManager(KafkaSpoutConfig<K, V> kafkaSpoutConfig, Fields fields)`

revans2 replied:
I'm a bit confused about what benefit that would give.  The Fields come from the SpoutConfig, by way of the RecordTranslator.  Why would you want to override the Fields but not the RecordTranslator that must conform to those fields?  I just don't see value it separating two things that are highly coupled together.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/trident/TridentKafkaState.java:
```diff
@@ -0,0 +1,114 @@
+    private TridentTupleToKafkaMapper mapper;
+    private KafkaTopicSelector topicSelector;
+
+    public TridentKafkaState withTridentTupleToKafkaMapper(TridentTupleToKafkaMapper mapper) {
```
Wouldn't it be preferable the class to be immutable ?

revans2 replied:
This is code that was "moved" like with the KafkaBolt for storm-kafka to storm-kafka-client.  If we really want to make it immutable we can, but I think that is beyond the scope of this JIRA

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/trident/TridentKafkaState.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    public void prepare(Properties options) {
+        if (mapper == null) throw new NullPointerException("mapper can not be null");
```
Objects.requireNonNull("mapper cannot be null");
..

revans2 replied:
Objects does not exist in java 6 and I would prefer to keep the code compatible as mush as possible to avoid extra rework when pulling these changes back.  If you insist I will do it.

srdo replied:
I thought 1.x was targeting at least Java 7?

revans2 replied:
@srdo you are right.

```
                <artifactId>maven-compiler-plugin</artifactId>
                <configuration>
                    <source>1.7</source>
                    <target>1.7</target>
                </configuration>
```
I'll update it.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/selector/KafkaTopicSelector.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.Serializable;
+
+public interface KafkaTopicSelector extends Serializable {
+    String getTopic(Tuple tuple);
```
since in some scenarios it's OK that this value returns null, wouldn't it be much clearer to return `Optional<String>` instead?

revans2 replied:
File another JIRA and we can look into it.  I think it is beyond the scope of this one.

HeartSaVioR replied:
IMHO we need to try avoiding JDK 8 feature for public API unless we decided to release minor version only for 2.0.0. But anyway I also think this is beyond of the scope.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/trident/mapper/FieldNameBasedTupleToKafkaMapper.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public K getKeyFromTuple(TridentTuple tuple) {
+        return (K) tuple.getValueByField(keyFieldName);
```
ClassCastException?

revans2 replied:
It is possible but that is the point of the generics.  To try and reduce the likelihood of it happening.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/trident/mapper/FieldNameBasedTupleToKafkaMapper.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Override
+    public V getMessageFromTuple(TridentTuple tuple) {
```
ClassCastException?

revans2 replied:
It is possible but that is the point of the generics. To try and reduce the likelihood of it happening.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:
```diff
@@ -225,11 +213,7 @@ public void nextTuple() {
-                    } catch (RetriableException e) {
-                        LOG.error("Failed to poll from kafka.", e);
-                    }
+                    setWaitingToEmit(pollKafkaBroker());
```
That's right it was a merge error.  Great catch though.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -61,129 +66,244 @@
+            this.maxRetries = builder.maxRetries;
+            this.firstPollOffsetStrategy = builder.firstPollOffsetStrategy;
+            this.maxUncommittedOffsets = builder.maxUncommittedOffsets;
+            //this could result in a lot of class case exceptions at runtime,
```
Sorry the comment was not updated.  I will do that.  Originally it was for API compatibility, but now it is to not force someone to reset the translator when they change a key or value type.  Specifically the default translator that should work on any types.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -61,129 +66,244 @@
-        public Builder<K,V> setKeyDeserializer(SerializableDeserializer<K> keyDeserializer) {
-            this.keyDeserializer = keyDeserializer;
-            return this;
+        public <NK> Builder<NK,V> setKey(SerializableDeserializer<NK> keyDeserializer) {
```
Because we are changing the key type of the deserializer.

srdo replied:
I think what I was asking was "Why would you need to create a Builder with key type A and then later set a deserializer for keys of type B"? :)

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/selector/FieldIndexTopicSelector.java:
```diff
@@ -0,0 +1,52 @@
+ * Uses field with a given index to select the topic name from a tuple .
+ */
+public class FieldIndexTopicSelector implements KafkaTopicSelector {
+    private static final long serialVersionUID = -3830575380208166367L;
```
Because I missed them.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/selector/FieldNameTopicSelector.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public String getTopic(Tuple tuple) {
```
I agree but both for backwards compatibility and to limit the scope of this JIRA I would rather see it done in a follow on JIRA

revans2 replied:
I filed https://issues.apache.org/jira/browse/STORM-2318 for this

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/ByTopicRecordTranslator.java:
```diff
@@ -0,0 +1,93 @@
+        for (String stream : translator.streams()) {
+            Fields fromTrans = translator.getFieldsFor(stream);
+            Fields cached = streamToFields.get(stream);
+            if (cached != null && !fromTrans.equals(cached)) {
```
```
ByTopicRecordTranslator<A,B> trans = new ByTopicRecordTranslator((rec) -> Arrays.asList(rec.offset()), new Fields("offset"), "default");
trans.forTopic("specialTopic", (rec) -> Arrays.asList(rec.offset(), rec.message()), new Fields("offset", "message"), "default");
```
At this point we have tried to declare that the "default" stream has Fields ["offset"], and ["offset", "message"]  This is not supported by storm so we should not allow anyone to configure the spout to do this.

streamToFields is not yet updated for the new translator we are adding it yet.  We do it after we have verified that the Fields match for anything we have done already.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/ByTopicRecordTranslator.java:
```diff
@@ -0,0 +1,93 @@
+            Fields fromTrans = translator.getFieldsFor(stream);
+            Fields cached = streamToFields.get(stream);
+            if (cached != null && !fromTrans.equals(cached)) {
+                throw new IllegalArgumentException("Stream " + stream + " currently has Fields of " + cached + " which is not the same as those being added in " + fromTrans);
```
In this case it is the argument being passed in that is bad and we are rejecting it.  They could come back and switch it to a new stream which would be fine.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/Subscription.java:
```diff
@@ -0,0 +1,62 @@
+     * @param listener the rebalance listener to include in the subscription
+     * @deprecated please use the version with the TopologyContext in it
+     */
+    public <K, V> void subscribe(KafkaConsumer<K, V> consumer, ConsumerRebalanceListener listener) {
```
I didn't feel good about deprecating it in the first place and will remove it.  It just means that I will also pull in the changes I made for STORM-2236 back to 1.1 too.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:
```diff
@@ -325,15 +310,19 @@ private boolean emitTupleIfNotEmitted(ConsumerRecord<K, V> record) {
-                final List<Object> tuple = tuplesBuilder.buildTuple(record);
-                kafkaSpoutStreams.emit(collector, tuple, msgId);
+                final List<Object> tuple = kafkaSpoutConfig.getTranslator().apply(record);
+                if (tuple instanceof KafkaTuple) {
```
The javadocs for RecordTranslator state.
```
     * @return the objects in the tuple.  Return a {@link KafkaTuple}
     * if you want to route the tuple to a non-default stream
```

All of the provided implementations support this.  Both Simple and ByTopic, by way of the SimpleRecordTranslator.

I will add a section in the docs to talk about this, but it only really matters if you are writing your own record translator from scratch, instead of using the built in ones that should cover the vast majority of the use cases.

revans2 replied:
Yes, and it is documented in RecordTranslator.  I will add in more documentation on it though.

The reason specifically for this was because the spout is not able to keep track of a single message being emitted to multiple streams.  It would get confused and ack it before it was truly done.  This makes it impossible for that to happen.  What is more the built in record translators should cover 99% of the use cases, so the fact that it is not super well documented should be more of a corner case.

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/TopicPartitionComparator.java:
```diff
@@ -1,16 +1,35 @@
-            return o1.partition() - o2.partition();
-        }
-    }
+	public static final TopicPartitionComparator INSTANCE = new TopicPartitionComparator();
```
Enums for things that are not an enumeration feels a little weird. Wouldn't you get the same result by just adding a private no-args constructor?

revans2 replied:
I'm not sure we need to guarantee it.  The old code had the INSTANCE where it was used, I moved it here in hopes that others might use it.  I could make the constructor private if we really want it to be a singleton, but I don't think it is a requirement.

hmcl replied:
@srdo that is not true. See Effective Java Item 3. Yes, private constructor would work.

@revans2 either way is fine. My comment was motivated because I believe that either we should make it a singleton or not. If not, let the users simple instantiate it with new and that's it. I am a believer in consistency.

revans2 replied:
@hmcl I agree I went half way to making it a true singleton and I should finish the job.

srdo replied:
@hmcl Neat, hadn't considered those arguments for using enum.

srdo commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/RoundRobinManualPartitioner.java:
```diff
@@ -15,14 +15,26 @@
+		int thisTaskIndex = context.getThisTaskIndex();
+		int totalTaskCount = context.getComponentTasks(context.getThisComponentId()).size();
+		Set<TopicPartition> myPartitions = new HashSet<>(allPartitions.size()/totalTaskCount+1);
+		for (int i = thisTaskIndex; i < allPartitions.size(); i += totalTaskCount) {
```
I don't think so. Either i is in [0, allPartitions.size()-1] or the loop is over.

revans2 replied:
No.  `i < allPartitions.size()` guarantees that we will never call get on allPartitions with an index that is out of bounds.  The Set is just setting the initial size to avoid more memory allocation.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/NamedSubscription.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @Override
+    public String getTopicsString() {
```
It is in the parent javadocs.  That is around the override

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaTuple.java:
```diff
@@ -15,22 +15,33 @@
+        super(vals);
+    }
+    
+    public KafkaTuple routedTo(String stream) {
```
This is because the constructor is varadic following the Values parent class. It is ambiguous to have

```
    public KafkaTuple(Object... vals)
```

and any other constructor.  Java complains.

revans2 commented:
@hmcl I think I addressed all of your review comments.  I also re-based to deal with the merge conflict. If everything looks good I will squash some of the commits.  There are a lot.

revans2 commented:
The travis test failures are unrelated.

revans2 commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:
```diff
@@ -447,7 +440,10 @@ private void shutdown() {
     @Override
     public void declareOutputFields(OutputFieldsDeclarer declarer) {
-        kafkaSpoutStreams.declareOutputFields(declarer);
+        RecordTranslator<K, V> translator = kafkaSpoutConfig.getTranslator();
```
I agree I think we need a fairly standard way to Tuple->X and X->Tuple.  But really it would just end up looking like

```java
public interface <X> TupleToX {
    public X fromTuple(Tuple t);
}

public interface X XToTuple {
    public Tuple toTuple(X record);
}
```

But I don't know how much code reuse there could be, because X is so different for each library we are going to interface with.

srdo commented:
There are still a few files with tabs mixed in to spaces, so autoformatting all the files might be good. I also still don't really understand why the KafkaSpoutConfig builder needs to support setting a deserializer with a different generic type to the one the builder was instantiated with.

+1 though.

ptgoetz commented:
+1

hmcl commented:
+1

My only suggestion at this point is that we file an umbrella JIRA covering some of the follow up comments that haven't been addressed, such as minor refactoring, improved (java) docs, immutability, etc. Basically to just capture the most meaningful comments that makes sense handle later.

Thanks @revans2 for all your patience addressing the many review comments

revans2 commented:
Everyone,

Sorry this took so long.  I kept getting pulled off on critical issues.  I have rebased/squashed everything (except the code revert) and also fixed the issue with tabs vs spaces.  Since it was just white space issues I don't think I need more +1s but if someone wants to take a look again please feel free to.  I will do the same for the 1.x version, and then file the umbrella jira that @hmcl requested.

hmcl commented:
Thanks @revans2. Can you please just amend the commit log [messages](https://github.com/apache/storm/pull/1808/commits) as I believe they do not really reflect the changes in this patch. One of the messages has STORM-1997 title, and the other has a cryptic merge log message.

Also, can you please do the same for 1.x-branch [patch](https://github.com/apache/storm/pull/1868)

revans2 commented:
@hmcl github like to truncate long messages.

```
$ git log STORM-2225
commit c9f9348e3de0917020a8c903f4c9d026b7dc6204
Author: Robert (Bobby) Evans <evans@yahoo-inc.com>
Date:   Tue Nov 29 21:39:26 2016 -0600

    STORM-1997: copy state/bolt from storm-kafka to storm-kafka-client
    STORM-2225: change spout config to be simpler.
    STORM-2228: removed ability to request a single topic go to multiple streams
    STORM-2236: Reimplemented manual partition management on top of STORM-2225

commit 74a77aa6494957b941df55d66ff64b6561eee4a8
Author: Robert (Bobby) Evans <evans@yahoo-inc.com>
Date:   Wed Jan 18 10:34:30 2017 -0600

    Revert "Merge branch 'STORM-2236' of https://github.com/MediaV/storm into STORM-2236"

    This reverts commit e2f87c304297ce95f01708e0934c44ee385fa0bb, reversing
    changes made to 00fed518d14b485e2a9e9cdd91ade3cc9d68dc8d.

```

If you really want me to squash the code change with the revert and only reference a single JIRA in the commit messages I can.  But I thought it would be better if I am addressing 4 separate (but very closely related) JIRA I should keep them all together.

hmcl commented:
@revans2 this is perfect. These 4 commits make it much easier to track down. I confused the way GitHub truncates the messages - my bad.

+1

HeartSaVioR commented:
Thanks for the great work and patience. +1
